CentOS/duffy
# Add tasks module and `worker` subcommand

The tasks module will contain potentially long-running operations, like checking node pools and topping them up. These tasks will be performed by a Celery worker in the background and only triggered from the controller function in order to not block the API.

Additionally, add the `postgresql` extra to the package which will install the `psycopg2` Python package as a dependency for production instances.

# How to test

- Ensure that testing configuration has a `celery` block like what was added ti `etc/duffy-example-config.yaml`.
- Run the worker (all arguments after `worker` are passed to the celery CLI), e.g.:
```
(duffy) nils@makake:~/src/fedora-infra/duffy (dev--tasks)> duffy -c etc/duffy-config.yaml worker -E
 
 -------------- celery@makake v5.2.1 (dawn-chorus)
--- ***** ----- 
-- ******* ---- Linux-5.15.6-200.fc35.x86_64-x86_64-with-glibc2.34 2021-12-08 16:18:05
- *** --- * --- 
- ** ---------- [config]
- ** ---------- .> app:         duffy.tasks:0x7f18d0bfbc70
- ** ---------- .> transport:   redis://localhost:6379//
- ** ---------- .> results:     redis://localhost:6379/
- *** --- * --- .> concurrency: 12 (prefork)
-- ******* ---- .> task events: ON
--- ***** ----- 
 -------------- [queues]
                .> celery           exchange=celery(direct) key=celery
```
- Trigger a task in a different shell:
```
(duffy) nils@makake:~/src/fedora-infra/duffy (dev--tasks)> celery -b redis:// --result-backend redis:// call duffy.tasks.check_pools
8400dd65-516a-4535-b2c9-07e3129ed83d
```
- Check the output in the running worker:
```
  -------------- celery@makake v5.2.1 (dawn-chorus)
--- ***** ----- 
-- ******* ---- Linux-5.15.6-200.fc35.x86_64-x86_64-with-glibc2.34 2021-12-08 16:18:05
- *** --- * --- 
- ** ---------- [config]
- ** ---------- .> app:         duffy.tasks:0x7f18d0bfbc70
- ** ---------- .> transport:   redis://localhost:6379//
- ** ---------- .> results:     redis://localhost:6379/
- *** --- * --- .> concurrency: 12 (prefork)
-- ******* ---- .> task events: ON
--- ***** ----- 
 -------------- [queues]
                .> celery           exchange=celery(direct) key=celery
                

[2021-12-08 16:18:19,031: WARNING/ForkPoolWorker-7] Checking pools...
```
